File: app/src/main/java/com/spacex/launches/spacexrocketlaunches/rocketList/RocketListAdapter.java
package com.spacex.launches.spacexrocketlaunches.rocketList;

import android.view.LayoutInflater;
import android.view.ViewGroup;
import androidx.annotation.NonNull;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.RecyclerView;
import com.spacex.launches.spacexrocketlaunches.databinding.RocketVhBinding;
import com.spacex.launches.spacexrocketlaunches.rocketList.RocketVH.OnRocketClickListener;
import java.util.List;

public class RocketListAdapter extends RecyclerView.Adapter<RocketVH> {

  private final OnRocketClickListener onRocketClickListener;
  private List<RocketUIM> rockets;

  RocketListAdapter(
      @NonNull final List<RocketUIM> rockets,
      @NonNull final OnRocketClickListener onRocketClickListener) {
    this.onRocketClickListener = onRocketClickListener;
    this.rockets = rockets;
  }

  void refreshData(@NonNull final List<RocketUIM> rockets) {
    DiffUtil.calculateDiff(new RocketsDiffCallback(this.rockets, rockets)).dispatchUpdatesTo(this);
    this.rockets = rockets;
  }

  @NonNull
  @Override
  public RocketVH onCreateViewHolder(@NonNull final ViewGroup parent, final int viewType) {
    return new RocketVH(
        RocketVhBinding.inflate(LayoutInflater.from(parent.getContext()), parent, false),
        onRocketClickListener);
  }

  @Override
  public void onBindViewHolder(@NonNull final RocketVH holder, final int position) {
    holder.onBind(rockets.get(position));
  }

  @Override
  public int getItemCount() {
    return rockets.size();
  }
}

File: app/src/main/kotlin/com/spacex/ui/rocketList/RocketListAdapter.kt
package com.spacex.ui.rocketList

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.RecyclerView
import com.spacex.ui.databinding.RocketVhBinding
import com.spacex.ui.rocketList.RocketVH.OnRocketClickListener

internal class RocketListAdapter(
        private var rockets: List<RocketUIM>?,
        private val onRocketClickListener: OnRocketClickListener) : RecyclerView.Adapter<RocketVH>() {

    fun refreshData(rockets: List<RocketUIM>) {
        DiffUtil.calculateDiff(RocketsDiffCallback(this.rockets!!, rockets)).dispatchUpdatesTo(this)
        this.rockets = rockets
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RocketVH {
        return RocketVH(
                RocketVhBinding.inflate(LayoutInflater.from(parent.context), parent, false),
                onRocketClickListener)
    }

    override fun onBindViewHolder(holder: RocketVH, position: Int) {
        holder.onBind(rockets!![position])
    }

    override fun getItemCount(): Int {
        return rockets!!.size
    }
}
